File: lingting-core/src/main/java/live/lingting/framework/api/ScrollCursor.java
package live.lingting.framework.api;

import live.lingting.framework.function.ThrowingFunction;
import live.lingting.framework.util.CollectionUtils;
import live.lingting.framework.value.CursorValue;

import java.util.List;

/**
 * @author lingting 2023-12-29 11:32
 */
public class ScrollCursor<T, S> extends CursorValue<T> {

	private final ThrowingFunction<S, ScrollResult<T, S>> scroll;

	private S scrollId;

	public ScrollCursor(ThrowingFunction<S, ScrollResult<T, S>> scroll, S scrollId, List<T> data) {
		this.scroll = scroll;
		this.scrollId = scrollId;
		// 初始数据就为空, 直接结束
		if (CollectionUtils.isEmpty(data)) {
			empty = true;
		}
		else {
			current.addAll(data);
		}
	}


	@Override
	protected List<T> nextBatchData() {
		ScrollResult<T, S> result = scroll.apply(scrollId);
		scrollId = result.getCursor();
		return result.getRecords();
	}

}

File: lingting-core/src/main/kotlin/live/lingting/framework/api/ScrollCursor.kt
package live.lingting.framework.api

import live.lingting.framework.function.ThrowingFunction
import live.lingting.framework.value.CursorValue

/**
 * @author lingting 2023-12-29 11:32
 */
class ScrollCursor<T, S>(
    params: ScrollParams<S>,
    private var function: ThrowingFunction<ScrollParams<S>, ScrollResult<T, S>>,
) : CursorValue<T>() {

    private var scrollId = params.cursor

    private val size = params.size

    override fun nextBatchData(): List<T> {
        val params = ScrollParams(size, scrollId)
        val result = function.apply(params)
        scrollId = result.cursor
        return result.records
    }

}
